specs/granite/header_accumulator.md

```diff
@@ -0,0 +1,302 @@
+produced digest as a big-endian 32-byte unsigned integer.
+
+| Byte Range | Description      |
+| ---------- | ---------------- |
+| `[0, 20)`  | Truncated Digest |
+| `[20, 32)` | Discarded        |
+
+## Block Execution Changes
+
+After Granite activation, every time `block.number % HEADER_BATCH_SIZE == 0`, the execution layer will:
+
+1. Craft a binary merkle tree of depth `HEADER_BATCH_TREE_DEPTH` with the block hashes of
+   `[block.number - HEADER_BATCH_TREE_DEPTH, block)` and compute its root, using the `keccak256` hash function.
+1. Fetch the previous merkle stack from `block.number - HEADER_BATCH_TREE_DEPTH + 1`'s Header.
+   - If the current block is the first header accumulation block, the execution client will begin with an empty
+     merkle stack consisting of zero hashes from the leaf node to the root node, set the `header_batch_num` to `0`,
+     and continue to the next step.
+   - If the current block is not the first header accumulation block, and the block header at
+     `block.number - HEADER_BATCH_TREE_DEPTH` does not contain a valid header accumulator `extraData` field,
+     block execution should _always_ fail immediately.
+   - If the current block is not the first header accumulation block, and the block header at
+     `block.number - HEADER_BATCH_TREE_DEPTH` does contain a valid header accumulator `extraData` field, continue to
+     the next step.
+1. Insert the root of the header batch tree computed in step #1 into the merkle stack retrieved from the previous
+   header accumulation block, using the insertion function defined below.
+1. Compute the new accumulator tree root with the modified stack.
+1. Encode the merkle root and modified merkle stack as defined below, and insert into the executed block's header
+   prior to sealing it.
+
+### Block Validity Changes
+
+1. If `block.number % HEADER_BATCH_SIZE == 0`, the accumulator tree merkle root in the `extraData` must
+   contain all previous header batches from the previous accumulator tree, in-order, with the current header
+   batch appended in the next available leaf.
+   - If `block.number` is the first header accumulation block, the accumulator merkle root in `extraData` must contain
+     only the current header batch at the `0`th index.
+1. If `block.number % HEADER_BATCH_SIZE == 0`, the `header_batch_num` in the header must equal the
+   previous header accumulation block's `header_batch_num` plus `1`.
+   - If `block.number` is the first header accumulation block, `header_batch_num` must equal `1`.
+1. If `block.number % HEADER_BATCH_SIZE == 0`, the `extraData` size must be
+   `8 + MERKLE_STACK_SIZE`.
+1. If `block.number % HEADER_BATCH_SIZE > 0`, the `extraData` size must be `0`.
+
+### Header Changes
+
+#### `extraData` format
+
+For header accumulation blocks (`block.number % HEADER_BATCH_SIZE == 0`), the accumulator tree root
+as well as the merkle stack should be encoded as follows:
+
+```txt
```

Comment:
Curious reviewers' thoughts. Should we also append the header batch root to this? Small bit extra data, though makes it so that we wouldn't need to reconstruct the batch tree off-chain for finding leaves of the accumulator tree.

Reply:
What's the size of the redundant data being attached? And would we do any validation that the batch-root included matches what is calculated?

Reply:
If we did add this, it would be an extra 32 bytes per `HEADER_BATCH_SIZE` blocks in the `extraData` field. We would need to add a validation step, where we verify that the leaf (the header batch tree root) is verified to be included within the tree that the merkle stack represents.